examples/pruning/lightgbm_with_custom_cv_function.py

```diff
@@ -0,0 +1,404 @@
+# Functions
+
+"""
+This configurable Optuna example demonstrates the use of pruning when using
+custom cross-validation functions based on LightGBM classifiers or regressors.
+
+In this example, we optimize cross-validation accuracy of either a classification model
+of breast cancer probability or a regression model of Boston housing prices,
+with a custom CV function custom_cv_fun() applying LightGBM's .fit() method
+to each cross-validation fold separately, instead of using the standard
+built-in .cv() function.
+
+Throughout training, an Optuna pruner observes intermediate results and if necessary
+stops early unpromising trials using a callback, resulting in noticeably faster execution
+times with unaffected accuracy.
+
+The example emphasizes reproducibility, giving the user precise control over all seed
+values that are being used to let her arrive at the same results over repeated studies.
+
+You can run this example using python 3 interpreter, e.g. as follows:
+    $ python lightgbm_with_custom_cv_function.py
+"""
+
+# Import packages
+
+from datetime import datetime
+from pprint import pprint
+
+from lightgbm import LGBMClassifier
+from lightgbm import LGBMRegressor
+import numpy as np
+import pandas as pd
+from sklearn import datasets
+from sklearn.model_selection import KFold
+from sklearn.model_selection import train_test_split
+
+import optuna
+from optuna.pruners import MedianPruner
+from optuna.samplers import TPESampler
+
+
+# Custom functions
+
+
+def custom_cv_fun(
+    lgbm_params,
+    X,
+    y,
+    objective="binary",
+    eval_metric="auc",
+    eval_name="valid",
+    n_estimators=100,
+    early_stopping_rounds=10,
+    nfold=5,
+    random_state=123,
+    callbacks=[],
+    verbose=False,
+):
+
+    # create placeholders for results
+    fold_best_iterations = []
+    fold_best_scores = []
+
+    # get feature names
+    feature_names = list(X.columns)
+
+    # split data into k folds
+    kf = KFold(n_splits=nfold, shuffle=True, random_state=random_state)
+    splits = kf.split(X, y)
+
+    # iterate over folds
+    for n, (train_index, valid_index) in enumerate(splits):
```

Comment:
`n` is not used and we can remove it.
```suggestion
    for train_index, valid_index in splits:
```